cmd/clusterctl/pkg/client/config/providers_client.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/klog"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+)
+
+const (
+	CAPI               = "cluster-api"
+	ProvidersConfigKey = "providers"
+)
+
+// ProvidersClient has methods to work with provider configurations.
+type ProvidersClient interface {
+	// List returns all the provider configurations, including provider configurations hard-coded in clusterctl
+	// and user-defined provider configurations read from the clusterctl configuration file.
+	// In case of conflict, user-defined provider override the hard-coded configurations.
+	List() ([]Provider, error)
+
+	// Get return the configuration for the provider with a given name.
+	// In case the name does not correspond to any existing provider, an error is returned.
+	Get(name string) (*Provider, error)
+}
+
+// providersClient implements ProvidersClient.
+type providersClient struct {
+	reader Reader
+}
+
+// ensure providersClient implements ProvidersClient.
+var _ ProvidersClient = &providersClient{}
+
+func newProvidersClient(reader Reader) *providersClient {
+	return &providersClient{
+		reader: reader,
+	}
+}
+
+func (p *providersClient) defaults() []Provider {
+
+	// clusterctl includes an hard-coded list of "default" providers in order to provide to the users the simplest
+	// out-of-box experience. Other providers can be added by using the clusterctl configuration file.
+
+	// if you are a developer of a community maintained provider, you can send a PR to extend the following list.
```

Comment:
can this face the fate of our CNI section of the kubeadm setup tutorial about CNI?
random (or any) provider want's to be included in the list of "popular providers".

Reply:
Possibly, but I would hope that the metadata exposed would prevent it from being included if the metadata falls out of date for a give cluster-api release (i.e. does not support v1alpha3 yet), and we can also provide some type of guidance on what is required to be included (timeframe to support a new cluster-api release, some type of CI automation signal, etc).

Reply:
as soon as https://github.com/kubernetes-sigs/cluster-api/issues/1802 get defined, I'm +1 to define some CI signal around this.